File: Top_Interview_150/Stack/python/150.py
from typing import List


class Solution:

    @staticmethod
    def calculate(num1: int, num2: int, operation: str) -> int:
        match operation:
            case "+":
                return num1 + num2
            case "-":
                return num1 - num2
            case "*":
                return num1 * num2
            case "/":
                return int(float(num1) / num2)
            case _:
                return 0

    def evalRPN(self, tokens: List[str]) -> int:
        stack = list()
        for token in tokens:
            if token.isdigit() or token[1:].isdigit():
                stack.append(int(token))
            else:
                x2 = stack.pop()
                x1 = stack.pop()
                stack.append(self.calculate(x1, x2, token))

        return stack[0]


solution = Solution()
print(solution.evalRPN(["-78", "-33", "196", "+", "-19", "-", "115", "+", "-", "-99", "/", "-18", "8", "*", "-86", "-", "-", "16", "/", "26", "-14", "-", "-", "47", "-", "101", "-", "163", "*", "143", "-", "0", "-", "171", "+", "120", "*", "-60", "+", "156", "/", "173", "/", "-24", "11", "+", "21", "/", "*", "44",
      "*", "180", "70", "-40", "-", "*", "86", "132", "-84", "+", "*", "-", "38", "/", "/", "21", "28", "/", "+", "83", "/", "-31", "156", "-", "+", "28", "/", "95", "-", "120", "+", "8", "*", "90", "-", "-94", "*", "-73", "/", "-62", "/", "93", "*", "196", "-", "-59", "+", "187", "-", "143", "/", "-79", "-89", "+", "-"]))

File: Top_Interview_150/Stack/cpp/150.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>

using namespace std;

class Solution
{
public:
    int calculate(int num1, int num2, string operation)
    {
        if (operation == "+")
            return num1 + num2;
        else if (operation == "-")
            return num1 - num2;
        else if (operation == "*")
            return num1 * num2;
        else if (operation == "/")
            return (int)((double)num1 / num2);
        else
            return 0;
    }

    int evalRPN(vector<string> &tokens)
    {
        stack<int> stack;
        for (string &token : tokens)
        {
            if (isdigit(token[0]) || (token.length() > 1 && isdigit(token[1])))
                stack.push(stoi(token));
            else
            {
                int x2 = stack.top();
                stack.pop();
                int x1 = stack.top();
                stack.pop();
                stack.push(calculate(x1, x2, token));
            }
        }

        return stack.top();
    }
};